docs(how-to guides): added documentation for maps creation (#140)
autowarefoundation/autoware-documentation: docs/how-to-guides/how-to-create-maps-for-autoware.md

# Map data for Autoware

Autoware relies on high-definition maps (HD maps), including both point cloud and vector maps, to perform various tasks such as localization, route planning, traffic light detection, and other traffic participants' trajectory prediction. In order to use these functionalities in Autoware, compatible HD maps of the driving environment need to be created.

## Creating HD maps

> You can use various software, both open-source and proprietary, to create point cloud and vector maps. However, you need to ensure that the output maps are compatible with Autoware. Specifications of point cloud and vector maps that Autoware requires are given below, along with examples of software you can use to create them.

Author: According to this comment as well as [this](https://github.com/autowarefoundation/autoware-documentation/pull/140#discussion_r894219185) and [this](https://github.com/autowarefoundation/autoware-documentation/pull/140#discussion_r894224085), I think I will rename the title and just remove the first section. I also thought it was a bit weird too but so I'm gladly changing this.

> ### Point cloud map

Reviewer: ```suggestion
### Creating a point cloud map
```

> A 3D point cloud map is primarily used for LiDAR-based localization in Autoware. In order to determine the current position and orientation of the vehicle, a live scan captured from LiDAR unit(s) is matched against a pre-generated 3D point cloud map. Therefore, an accurate point cloud map is crucial for good localization results. Specifications of an Autoware-compatible point cloud map are as follows:

Author: Noted!

- A map must cover the entire operational area of the vehicle and should include an additional buffer zone of at least 200 m in all directions.
- A map is saved using [PCD (Point Cloud Data) file format](https://pointclouds.org/documentation/tutorials/pcd_file_format.html).
- A map can be saved as a single PCD file or divided into multiple PCD files.
- Each point in a map must contain X, Y, and Z coordinates.
- An intensity or RGB value of each point can be included in the PCD file but is optional.
- A map must be smaller than 1 GB as it is limited by ROS message size.
- A map resolution should be at least 0.2 m to yield reliable localization results.
- Although a map can generally be in either local or global coordinates, it must be in global coordinates (georeferenced) if you want to use GNSS data for localization.

> - [Military Grid Reference System (MGRS)](https://en.wikipedia.org/wiki/Military_Grid_Reference_System) is used as a coordinate system for a georeferenced map. In a georeferenced map, the X and Y coordinates of each point represent the point's location within a 100 km MGRS grid, while the Z coordinate represents the point's elevation.

Reviewer 2: Should the map follow MGRS when using it for Autoware?

I think this is necessary when the readers want to use MGRS-based GNSS sensor input, but otherwise, IMO the map does not necessarily should follow the MGRS format. (correct me if I am wrong!)

So what do you think about removing this line? 
```suggestion
```

Author: Basically, if GNSS is not used it can be in local coordinates as mentioned in the prior bullet point. However, if users want to georeference the map it has to be in MGRS. I think it's important to note that because there are other projections that can be used, UTM, Rectangular plane. Especially, since Autoware.ai used to use Rectangular plane projection and now we switched to MGRS.

Therefore, if you're ok with it I would like to keep this as some users might find it useful.

Reviewer 2: Thanks.

I assumed that we can change the format of map coordinates (among MGRS, UTM, or plane) from [this parameter in `gnss_poser`](https://github.com/autowarefoundation/autoware.universe/blob/f807914b60e70fb1ad4a50bdd942bd2330518fba/sensing/gnss_poser/launch/gnss_poser.launch.xml#L15) when GNSS is available. 
If my understanding is correct, how about something like this?

```suggestion
- When GNSS is available, the coordination system of the map should be compatible with the GNSS data. In the current Autoware Universe, we support MGRS, UTM, and Rectangular plane format for the map (see [the launch file of gnss_pose](https://github.com/autowarefoundation/autoware.universe/blob/main/sensing/gnss_poser/launch/gnss_poser.launch.xml#L15) to set the parameter).
```

Author: I see. I think it's reasonable to include other coordinate systems as well. However, it also seems Lanelet2 loader also supports UTM actually (though it's not the [default choice](https://github.com/autowarefoundation/autoware.universe/blob/main/map/map_loader/src/lanelet2_map_loader/lanelet2_map_loader_node.cpp#L53)). 

I will try to reword the sentence a little bit but will reflect your idea of including other coordinate systems. Thank you!

> It is important to note that apart from [lidarslam-ros2](https://github.com/rsasaki0109/lidarslam_ros2), these open-source SLAM implementations are based on ROS 1. Therefore, it could be problematic if you want to install them on the same machine that runs Autoware, which is based on ROS 2. To avoid this problem, you can use Docker or install them on a different machine. Another problem is the ROSBAG version, as many of these SLAM implementations require ROSBAG 1 instead of ROSBAG 2 used by Autoware. For the ROSBAG version problem, you may use this [stand-alone converter](https://gitlab.com/MapIV/rosbags) to convert a ROSBAG 2 file to a ROSBAG 1 file and vice versa.

Author: Oh, I didn’t notice the ros2 branch. I will mention this in the text too. Also, thanks for the discussion link!

> It is important to note that apart from [lidarslam-ros2](https://github.com/rsasaki0109/lidarslam_ros2), these open-source SLAM implementations are based on ROS 1. Therefore, it could be problematic if you want to install them on the same machine that runs Autoware, which is based on ROS 2. To avoid this problem, you can use Docker or install them on a different machine. Another problem is the ROSBAG version, as many of these SLAM implementations require ROSBAG 1 instead of ROSBAG 2 used by Autoware. For the ROSBAG version problem, you may use this [stand-alone converter](https://gitlab.com/MapIV/rosbags) to convert a ROSBAG 2 file to a ROSBAG 1 file and vice versa.

Author: I think that’s because there are more ROS1-based SLAM implementations that are known to work well and do not strictly require other sensors except LiDAR (at least for now). Also, the ones that use ROS2 don’t seem to be as mature as those that are based on ROS1.

However, I understand that many users may prefer an easy-to-use implementation just to get started quickly. I think maybe if I switch their position and talk about ROS2-based ones first that would help. Essentially, what @LalithVipulananthan  suggested.

> If you prefer proprietary software that is easy to use, you can try a fully automatic mapping tool from [MAP IV, Inc.](https://www.map4.jp/), [_MapIV Engine_](https://www.map4.jp/map4_engine_en/). They currently provide a trial license for Autoware users free of charge.

Reviewer: The Japanese version is [here](https://www.map4.jp/technology#mapiv_engine), but I cannot seem to find an English version

> If you prefer proprietary software that is easy to use, you can try a fully automatic mapping tool from [MAP IV, Inc.](https://www.map4.jp/), [_MapIV Engine_](https://www.map4.jp/map4_engine_en/). They currently provide a trial license for Autoware users free of charge.

Author: It's not outdated but rather not ready yet. I requested the web team to create a landing page with that URL but it seems like it's not available just yet.

> ### Vector map (Road network map)

Reviewer 3: In order to get vector map, the best solution might be using vector map builder. 
IMO, it would be nice to get that answer a little sooner, since at the moment that information is at the end of the section.

Author: The reason I didn’t mention VMB first is that I thought it was not free for the public (I’ve always used it with a TeirIV e-mail). I felt like I should try to cover free and open-source options as much as possible before mentioning any proprietary tools. 

Since it’s free as long (as you create an account with a valid e-mail), I will switch the order and mention VMB first.

> ### Vector map (Road network map)

Reviewer: ```suggestion
### Creating a vector map
```

> Another type of map used in Autoware is a vector map. A vector map or road network map is required for route planning, traffic light detection, and other traffic participants' trajectory prediction. Specifications of an Autoware-compatible road network map are as follows:

Reviewer 3: I'm not sure but the word "vector map" and "road net map" sound like a different meaning. The latter is more rough map? 
IMO, it should unify the word  "vector map" if "road net map" is not commonly used.

Author: I think “road network map” is actually a more common term compared to “vector map” when people refer to a high-definition map that has details of lanes, road boundaries, junctions, etc.

As far as I understand, “vector map” is a format name. It is a proprietary format from AISAN used in [Autoware.ai](http://autoware.ai/) (ROS1 Autoware) to represent a road network map. In the current Autoware (ROS2 Autoware), the chosen format is Lanelet2. 

I think the term “vector map” is equivalent to “Lanelet2”, “OpenDrive”, or “OpenStreetMap”. They are all formats used to describe a road network.

Because of this, I thought it would be more clear to just use both words. However, if it seems to cause confusion here, maybe I can modify the text a little bit to make it clearer.

> Another type of map used in Autoware is a vector map. A vector map or road network map is required for route planning, traffic light detection, and other traffic participants' trajectory prediction. Specifications of an Autoware-compatible road network map are as follows:

Author: As I [replied](https://github.com/autowarefoundation/autoware-documentation/pull/140#discussion_r895357538) to the comment above, I think a road network map is a broader term and vector map actually refers to a format.

However, I think it's reasonable to keep the terminology consistent within the documentation (it is specifically aimed for Autoware users to understand after all). I think adding a sentence saying how vector map and road network map are related is a good idea.

However, the easiest way to create a complete road network map that is compatible with Autoware is to use [Vector Map Builder](https://tools.tier4.jp/vector_map_builder/), a web-based tool provided by [TIER IV, Inc.](https://www.tier4.jp/) Vector Map Builder allows you to directly create lanes and add additional regulatory elements such as stop signs or traffic lights in the correct format required by Autoware while using a point cloud map as a reference.

## Autoware-compatible map providers

Apart from creating HD maps yourself, you can also use a mapping service from the following Autoware-compatible map providers:

- [MAP IV, Inc.](https://www.map4.jp/)

> - [AISAN TECHNOLOGY CO., LTD.](https://www.aisantec.co.jp/)

Author: I think both of them do ( at least Map IV does), but let me check again.

# How to create maps for Autoware

Autoware relies on high-definition maps (HD maps), both point cloud maps and vector maps, to perform various tasks such as localization, route planning, traffic light detection, and predicting the trajectories of other vehicles and pedestrians. In order to use these functionalities in Autoware, compatible HD maps of the driving environment need to be created.

> To create point cloud and vector maps you can use various software, both open-source and proprietary. However, you need to ensure that the created maps are compatible with Autoware. Specifications of point cloud and vector maps that Autoware requires are given below, along with examples of software you can use to create them.

Author: I think the revised sentences look good to me. I'll add this change.

### Creating a point cloud map

A 3D point cloud map is primarily used for LiDAR-based localization in Autoware. In order to determine the current position and orientation of the vehicle, a live scan captured from one or more LiDAR units is matched against a pre-generated 3D point cloud map. Therefore, an accurate point cloud map is crucial for good localization results.

> The specifications of an Autoware-compatible point cloud map are as follows:

Author: @LalithVipulananthan Just to confirm. Did you mean to use H3 here?

- A map must cover the entire operational area of the vehicle and should include an additional buffer zone of at least 200 m in all directions.
- A map is saved using [PCD (Point Cloud Data) file format](https://pointclouds.org/documentation/tutorials/pcd_file_format.html).
- A map can be saved as a single PCD file or divided into multiple PCD files.
- Each point in a map must contain X, Y, and Z coordinates.
- An intensity or RGB value of each point can be included in the PCD file but is optional.
- A map must be smaller than 1 GB, [as per the current ROS message size limit](https://github.com/ros/ros_comm/issues/902).
- A map resolution should be at least 0.2 m to yield reliable localization results.
- A map can be in either local or global coordinates, but must be in global coordinates (georeferenced) to use GNSS data for localization.
- [Military Grid Reference System (MGRS)](https://en.wikipedia.org/wiki/Military_Grid_Reference_System) is used as a coordinate system for a georeferenced map. In a georeferenced map, the X and Y coordinates of each point represent the point's location within a 100 km MGRS grid, while the Z coordinate represents the point's elevation.

> Traditionally, a Mobile Mapping System (MMS) is used in order to create highly accurate large-scale point cloud maps. However, since a MMS requires high-end sensors for precise positioning, its operational cost can be very expensive and may not be suitable for a relatively small driving environment. Alternatively, a Simultaneous Localization And Mapping (SLAM) algorithm can be used to create a point cloud map from recorded LiDAR scans. Commonly used open-source SLAM implementations include [hdl-graph-slam](https://github.com/koide3/hdl_graph_slam) (LiDAR, IMU\*, GNSS\*), [LeGO-LOAM](https://github.com/facontidavide/LeGO-LOAM-BOR) (LiDAR, IMU\*), [LeGO-LOAM-BOR](https://github.com/RobustFieldAutonomyLab/LeGO-LOAM) (LiDAR), [LIO-SAM](https://github.com/TixiaoShan/LIO-SAM) (LiDAR, IMU, GNSS), and [lidarslam-ros2](https://github.com/rsasaki0109/lidarslam_ros2) (LiDAR, IMU\*). The required sensor data for each algorithm is specified in the parentheses, where an asterisk (\*) indicates that such sensor data is optional. Most of these algorithms already have a built-in loop-closure and pose graph optimization. However, if the built-in, automatic loop-closure fails or does not work correctly, you can use [Interactive SLAM](https://github.com/SMRT-AIST/interactive_slam) to adjust and optimize a pose graph manually. For supported LiDAR models, please check the Github repository of each algorithm.

Author: @LalithVipulananthan My apologies. I modified this section quite heavily before I saw this suggestion. Could you please take a look at the current version again and see if it still needs a modification?

- A map is in [Lanelet2](https://github.com/fzi-forschungszentrum-informatik/Lanelet2) format with[additional modifications](https://github.com/autowarefoundation/autoware.universe/blob/main/map/lanelet2_extension/docs/lanelet2_format_extension.md).
- A map must contain the shape and position information of lanes, traffic lights, stop lines, crosswalks, parking spaces, and parking lots.
- Each lanelet must contain information regarding its right of way, speed limit, traffic direction, and associated traffic lights, stop lines, and traffic signs.
- Unless it is the end/beginning of a road, each lanelet must be correctly connected to its predecessor, successors, left neighbor, and right neighbor for routing planning to be working.

The easiest way to create a vector map that is compatible with Autoware is to use [Vector Map Builder](https://tools.tier4.jp/vector_map_builder/), a web-based tool provided by [TIER IV, Inc.](https://www.tier4.jp/) Vector Map Builder allows you to directly create lanes and add additional regulatory elements such as stop signs or traffic lights in the correct format required by Autoware while using a point cloud map as a reference.

> Technically, you can also use open-source software such as [JOSM](https://josm.openstreetmap.de/) to create Lanelet2 primitives, including point and linestring, and save the map as an OSM file. However, various modifications, e.g., adding mandatory tags and relations, must be done manually to make the map compatible with Autoware. This process can be very tedious and time-consuming. Another open-source option worth mentioning is [MapToolbox](https://github.com/autocore-ai/MapToolbox), a plugin for [Unity](https://unity.com/) designed to create a Lanelet2 map for Autoware.

Reviewer: - Since MapToolbox seems to be the better OSS option, I think we should mention that first and put JOSM at the bottom.
- I also simplified the description of JOSM, as I do not think we need to go into details about why JOSM is bad.

```suggestion
For open-source software options, [MapToolbox](https://github.com/autocore-ai/MapToolbox) is a plugin for [Unity](https://unity.com/) specifically designed to create Lanelet2 maps for Autoware.

Although [JOSM](https://josm.openstreetmap.de/) is another open-source tool that can be used to create Lanelet2 maps, be aware that a number of modifications must be done manually to make the map compatible with Autoware. This process can be tedious and time-consuming and so use of JOSM is not recommended.
```

## Vector maps

A vector map contains highly accurate information about a road network, lane geometry, and traffic lights. It is required for route planning, traffic light detection, and predicting the trajectories of other vehicles and pedestrians.

### Vector map specifications

- It must cover the entire operational area of the vehicle and should include an additional buffer zone of at least 200 m in all directions.
- It must be in [Lanelet2](https://github.com/fzi-forschungszentrum-informatik/Lanelet2) format, with [additional modifications required by Autoware](https://github.com/autowarefoundation/autoware.universe/blob/main/map/lanelet2_extension/docs/lanelet2_format_extension.md).
- It must contain the shape and position information of lanes, traffic lights, stop lines, crosswalks, parking spaces, and parking lots.
- Each lanelet in the map must contain information regarding its right of way, speed limit, traffic direction, associated traffic lights, stop lines, and traffic signs.
- Except at the beginning or end of a road, each lanelet in the map must be correctly connected to its predecessor, successors, left neighbor, and right neighbor.

### Creating a vector map

> The easiest way to create an Autoware-compatible vector map is to use [Vector Map Builder](https://tools.tier4.jp/vector_map_builder/), a free web-based tool provided by [TIER IV, Inc.](https://www.tier4.jp/).

Reviewer: Replaced the VMB link with one that works without logging into a T4 account.

```suggestion
The easiest way to create an Autoware-compatible vector map is to use [Vector Map Builder](https://tools.tier4.jp/feature/vector_map_builder_ll2/), a free web-based tool provided by [TIER IV, Inc.](https://www.tier4.jp/).
```

Author: Thank you for the suggestion!

## Autoware-compatible map providers

If it is not possible to create HD maps yourself, you can use a mapping service from the following Autoware-compatible map providers instead:

- [MAP IV, Inc.](https://www.map4.jp/)
- [AISAN TECHNOLOGY CO., LTD.](https://www.aisantec.co.jp/)
- [TomTom](https://www.tomtom.com/)

The table below shows each company's mapping technology and the types of HD maps they support.

| **Company**                                               | **Mapping technology** | **Available maps**          |
| --------------------------------------------------------- | ---------------------- | --------------------------- |
| [MAP IV, Inc.](https://www.map4.jp/)                      | SLAM                   | Point cloud and vector maps |
| [AISAN TECHNOLOGY CO., LTD.](https://www.aisantec.co.jp/) | MMS                    | Point cloud and vector maps |
| [TomTom](https://www.tomtom.com/)                         | MMS                    | Vector map\*                |

!!! note

    Maps provided by TomTom use their proprietary AutoStream format, not Lanelet2.

>     The open-source [AutostreamForAutoware tool](https://github.com/tomtom-international/AutoStreamForAutoware) can be used to convert an AutoStream map to a Lanelet2 map.

Reviewer: ```suggestion
    The open-source [AutoStreamForAutoware tool](https://github.com/tomtom-international/AutoStreamForAutoware) can be used to convert an AutoStream map to a Lanelet2 map.
```
